### DISCUSSION TITLE
E2E testing with Nhost and Hasura

### DISCUSSION THREAD

**Discussion by seyaobey-dev** [nhost/nhost#834] [Q&A]
We're using nhost as a backend for our mobile react-native app. We would like to setup a dedicated environment where we will be running our nightly E2E tests. Ideally we would like to hit the real hasura database and not just mock server responses. What advice or guidance can you give on how to setup nhost and hasura in such a context? (create more dedicated nhost apps (dev, qa, ...etc)? ) 
More specifically during tests, users will be created (e.g E2E testing of user login/signup flow) and also various data will be created). How should be approach clearing up those data after our tests to get a clean environment for the next run? (e.g setup scripts running before tests start?...)

**Comment by beepsoft** [comment]
I just wanted to ask something similar, but in broader terms:

What is the best practice to test an nhost application, ie. serverless functions?

Naturally we need integration tests because we don't want to mock the whole nhost stack (postgres, hasura, auth, storage).

My current solution is this:

1. Testing with jest, where the tests can be run with 3 options:

	a) Use the project's local nhost as is, assuming it is already running
	b) Start the project's local nhost and stop it once the tests finished
	c) Use any remote nhost endpoint (running in the cloud)
        (Could be a 4th option to create a brand new nhost app for testing by copying the project's nhost's files.)
	
2. I have a clearDatabase() call, which calls the Hasura /v2/query endpoint and runs a bunch of SQL TRUNCATE-s before each test, so that the necessary tables are reset to their initial state.

This works nicely as long as it is OK to clear the whole database.

The only problem I have is that this way I cannot test functions, which are time dependent. For example, there might be a job, which needs to be run 3 days from now. There's no way to tell the containers to jump ahead in time. The solution I see for this is a way to run postgres, hasura, auth, storage in docker and the functions and tests by jest in the same node instance. When functions run in the node as the tests I can do something like:

```js
  const future = moment().add(3, 'days').toDate()
  const origDate = Date
  const mockFn = jest
    .spyOn(global, 'Date')
    .mockImplementation(() =>  {
      return future
    });
  // Do something in the future
  ...
  
  // undo mock, travel back in time
  global.Date = origDate
```

For this I would need to connect the container modules with the functions. It could be done by taking the `.nhost/docker-compose.json`, edit it to do these reconnections/reconfigurations and run it this way before starting up jest. 

Running functions along with jest would also allow mocking other services, which are used by the functions, and are not nhost related.

It would be great if there was some support for this in the nhost-cli.
